### PR TITLE
feat: update jwt library

### LINKED ISSUE #641: Update python-jose dep
python-jose introduces breaking changes in its latest version. We're on 0.6.2, latest is 1.2.0, which probably fixes multiple bugs we've had.

### PR DESCRIPTION
**Pull request by jrconlin**
Newest library allows for better detection.

closes #641 

@bbangert r?

### CODE REVIEW

**Comment by bbangert** [comment]
None of web/base.py or web/webpush.pyp needs updating, nor its tests to handle this?

File: autopush/settings.py [hunk 1]
```diff
@@ -363,7 +358,8 @@ def parse_endpoint(self, token, version="v0", ckey_header=None):
                 raise InvalidTokenException("Corrupted push token")
             if not public_key:
                 raise InvalidTokenException("Invalid key data")
-            if not constant_time.bytes_eq(sha256(public_key).digest(),
+            decoded_key = base64url_decode(repad(public_key))
```
Comment (kitcambridge) [inline autopush/settings.py]: `base64url_decode` already calls `repad`. 😄

File: autopush/tests/test_endpoint.py [hunk 2]
```diff
@@ -869,19 +869,18 @@ def test_decipher_public_key(self):
...
-        (_, crypto) = self._gen_jwt(header, payload)
-        crypto_key = utils.base64url_decode(crypto)
-
-        spki_header = ('0V0\x10\x06\x04+\x81\x04p\x06\x08*'
-                       '\x86H\xce=\x03\x01\x07\x03B\x00\x04')
-        eq_(decipher_public_key(crypto_key), crypto_key)
-        eq_(decipher_public_key('\x04' + crypto_key), crypto_key)
-        eq_(decipher_public_key(spki_header + crypto_key), crypto_key)
+        (_, crypto_key) = self._gen_jwt(header, payload)
+        base_key = base64.urlsafe_b64decode(repad(crypto_key))[-64:]
```
Comment (kitcambridge) [inline autopush/tests/test_endpoint.py]: If you use `utils.base64url_decode`, you won't need to repad anything.

File: autopush/utils.py [hunk 3]
```diff
@@ -142,7 +142,11 @@ def extract_jwt(token, crypto_key):
     # on the JWT object. Vapid is a bit more creative in how it
     # stores data into a JWT and breaks expectations. We would have to
     # turn off most of the validation in order for it to be useful.
-    return json.loads(jws.verify(token, vk, algorithms=["ES256"]))
+    return jwt.decode(token, vk, options=dict(
+        verify_aud=False,
+        verify_sub=False,
+        verify_exp=False,
```
Comment (kitcambridge) [inline autopush/utils.py]: Could you explain a bit why we aren't validating the contents? ISTM we'd want to reject expired JWTs, and a mismatched audience would indicate they're meant for another push server?

File: autopush/utils.py [hunk 4]
```diff
@@ -118,6 +117,7 @@ def decipher_public_key(key_data):
 
     """
     # key data is actually a raw coordinate pair
+    key_data = base64.urlsafe_b64decode(repad(key_data))
```
Comment (kitcambridge) [inline autopush/utils.py]: Ibid.

File: autopush/utils.py (hunk 3 above)
Comment (jrconlin) [inline autopush/utils.py]: Put simply, we are validating the contents, just not via JWT. We have additional code checks which return descriptive error messages for the important bits. In addition, our use of sub isn't quite to spec with what jwt expects.

Comment (kitcambridge) [inline autopush/utils.py]: Makes sense. Thanks!

**Comment by pjenvey** [comment]
Does this actually bump us to the latest python-jose (there's no base-requirements.txt change)?

**Comment by jrconlin** [comment]
It does, but it's back compatible. I'll add the base req once I am
somewhere immobile.

File: autopush/settings.py [hunk 5]
```diff
@@ -359,13 +355,17 @@ def parse_endpoint(self, token, version="v0", ckey_header=None):
...
             raise InvalidTokenException("Corrupted push token")
         if version == 'v2':
+            if not auth_header:
+                raise VapidAuthException("Missing Authorization Header")
             if len(token) != 64:
                 raise InvalidTokenException("Corrupted push token")
             if not public_key:
                 raise InvalidTokenException("Invalid key data")
-            if not constant_time.bytes_eq(sha256(public_key).digest(),
+            decoded_key = base64url_decode(public_key)
```
Comment (pjenvey) [inline autopush/settings.py]: If we needed the TypeError guard before might we still need it now?

File: autopush/web/validation.py [hunk 6]
```diff
@@ -299,13 +306,15 @@ def token_prep(self, d):
...
         )
         return d
 
     def validate_auth(self, d):
         auth = d["headers"].get("authorization")
-        if not auth:
-            return
+        needs_auth = d["token_info"]["api_ver"] == "v2"
+        if not auth and not needs_auth:
+                return
```
Comment (pjenvey) [inline autopush/web/validation.py]: I'm surprised flake8 didn't complain about this 8 space indent

**Comment by pjenvey** [comment]
r+

File: autopush/tests/test_endpoint.py [hunk 7]
```diff
@@ -869,19 +873,18 @@ def test_decipher_public_key(self):
...
                    "sub": "mailto:admin@example.com"}
-        (_, crypto) = self._gen_jwt(header, payload)
-        crypto_key = utils.base64url_decode(crypto)
-
-        spki_header = ('0V0\x10\x06\x04+\x81\x04p\x06\x08*'
-                       '\x86H\xce=\x03\x01\x07\x03B\x00\x04')
-        eq_(decipher_public_key(crypto_key), crypto_key)
-        eq_(decipher_public_key('\x04' + crypto_key), crypto_key)
-        eq_(decipher_public_key(spki_header + crypto_key), crypto_key)
+        (_, crypto_key) = self._gen_jwt(header, payload)
```
Comment (bbangert) [inline autopush/tests/test_endpoint.py]: Parens not needed on the left side.

**Comment by bbangert** [comment]
r+

File: autopush/tests/test_endpoint.py (hunk 7 above)
Comment (jrconlin) [inline autopush/tests/test_endpoint.py]: There's more than just this one present. Will fix this along with others in a later PR.

**Pull request merged**